service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/dataproccluster/startup.sh

```diff
@@ -667,13 +672,198 @@ Group=${LOGIN_USER}
 EnvironmentFile=/etc/environment
 EnvironmentFile=/etc/default/jupyter
 WorkingDirectory=${USER_HOME_DIR}
-ExecStart=/bin/bash -c '/opt/conda/miniconda3/bin/jupyter lab'
+ExecStart=/bin/bash -c '/opt/conda/miniconda3/bin/jupyter notebook &>> ${USER_HOME_DIR}/jupyter_notebook.log'
 Restart=on-failure
 
 [Install]
 WantedBy=multi-user.target
 EOF
 
+############################
+# Install Software Framework
+############################
+
+# If the script executer has set the "software-framework" property to "hail", then install hail after dataproc optional components are installed
+
+readonly SOFTWARE_FRAMEWORK="$(get_metadata_value "instance/attributes/software-framework")"
+
+if [[ "${SOFTWARE_FRAMEWORK}" == "HAIL" ]]; then
+  emit "Installing Hail..."
+
+  # Create a script to install hail
+  cat << EOF >"${HAIL_SCRIPT_PATH}"
+#!${MINICONDA_BIN}python
+# This modified hail installation script installs the necessary hail packages and jupyter extensions, but does not install jupyter or set up a jupyter service as it already handled by dataproc's jupyter optional component.
+# See: https://storage.googleapis.com/hail-common/hailctl/dataproc/0.2.120/init_notebook.py
+import json
+import os
+import subprocess as sp
+import sys
+from subprocess import check_output
+
+assert sys.version_info > (3, 0), sys.version_info
+
+if sys.version_info >= (3, 7):
+    def safe_call(*args, **kwargs):
+        sp.run(args, capture_output=True, check=True, **kwargs)
+else:
+    def safe_call(*args, **kwargs):
+        try:
+            sp.check_output(args, stderr=sp.STDOUT, **kwargs)
+        except sp.CalledProcessError as e:
+            print(e.output).decode()
+            raise e
+
+
+def get_metadata(key):
+    return check_output(['/usr/share/google/get_metadata_value', 'attributes/{}'.format(key)]).decode()
+
+def mkdir_if_not_exists(path):
```

Comment:
This wrapper function feels low-value. Would recommend calling os.makedirs(...) directly.